README.md
# Dynaweb

Dynaweb is a light framework to help you make a cool dynamic website without driving you crazy with a ton of implementation constraints...

## Installation

```
git clone https://github.com/EddieBreeg/Dynaweb.git
cd Dynaweb
pip install .
```

## Usage

Available soon...

## Authors

[Eddie Breeg](https://github.com/EddieBreeg) 


CHANGELOG.md
# 0.0.1

Initial release!


tests/my_proj/run.ps1
python -m dynaweb run


tests/my_proj/views.py
from json import load, dump
from imghdr import what
from datetime import datetime

def index(request):
    if request['queries'] is None:
        return open("index.html").read()
    return str(request['queries'])

def formTest(request):
    if request['method'] == 'POST':
        users = load(open('users.json'))
        name = request['data']['name']
        age = int(request['data']['age'])
        users[name] = age
        with open('users.json', 'w') as out:
            dump(users, out, indent=4)
        return "/"

    else:
        return open('form.html').read()

def fileForm(request):
    if request['method'] == 'GET':
        return open('file-form.html').read()
    content = request['data']['file'][0]
    imageType = what(None, h=content)
    if imageType is None:
        return '/file-form?error=wrong-type'
    time = datetime.utcnow()
    name = '{0}_{1}_{2}_{3}_{4}_{5}.{6}'.format(
        time.day, time.month, time.year, time.hour, time.minute, time.second, imageType
    )
    with open('media/{0}'.format(name), 'wb') as output:
        output.write(content)
    return '/media/{0}'.format(name)

src/dynaweb.py
from os import sys
from pathlib import Path
from os import path, makedirs
from utils.Server import Server
from http.server import HTTPServer
import json

# getting the directory the module has been called from
workingDir = Path().absolute()

if sys.argv[1] == "new":
    # creation of a new project
    projName = sys.argv[2]
    fullPath = path.join(workingDir, projName)
    if path.exists(fullPath):
        raise FileExistsError("Target location already exists")
    makedirs(fullPath)
    # generating the default settings
    defaults = """{
    "port": 1337,
    "host": "localhost",
    "static_dirs": ["static"],
    "media_dirs": ["media"]
}
"""
    with open(path.join(fullPath, "settings.json"), "w") as settingsFile:
        settingsFile.write(defaults)

elif sys.argv[1] == "run":
    # running the server
    try:
        settings = json.load(open(path.join(workingDir, "settings.json")))
    except FileNotFoundError:
        print("Not in a project directory, aborting.")
        exit()

    # defining the address
    if (l := len(sys.argv)) == 2:
        addr = (settings['host'], settings['port'])
    elif l == 3:
        addr = (sys.argv[2], settings['port'])
    else:
        addr = (sys.argv[2], int(sys.argv[3]))
    
    # creating the server
    server = HTTPServer(addr, Server)
    print("Server started on {0}:{1}".format(
        addr[0],
        addr[1]
    ))
    server.serve_forever()


src/BasicServer.py
from os import sys
from pathlib import Path
from os import path, makedirs
from BasicHTTPServer.Server import *

global workingDir
workingDir = Path().absolute()

if sys.argv[1] == "new":
    projName = sys.argv[2]
    fullPath = path.join(workingDir, projName)
    if path.exists(fullPath):
        raise FileExistsError("Target location already exists")
    makedirs(fullPath)
    defaults = """{
    "port": 1337,
    "host": "localhost",
    "static_dirs": ["static"],
    "media_dirs": ["media"]
}
"""
    with open(path.join(fullPath, "settings.json"), "w") as settingsFile:
        settingsFile.write(defaults)

elif sys.argv[1] == "run":
    print(workingDir)
    try:
        addr = json.load(open(path.join(workingDir, "settings.json")))
    except FileNotFoundError:
        print("Not in a project directory, aborting.")
        exit()

    addr = (addr['host'], addr['port'])
    server = HTTPServer(addr, Server)
    print("Server started on {0}:{1}".format(
        addr[0],
        addr[1]
    ))
    server.serve_forever()


src/utils/parse.py
from inspect import getmembers
import json
from urllib.parse import parse_qs, unquote
import re
from os.path import join, dirname, basename, exists, isfile
import sys

def parseForm(content):
    '''Parses the POST request to get the data'''
    for k in (data:=parse_qs(content.decode('utf-8'))):
        data[k] = data[k][0]
    return data


def parseURL(request, workingDir):     
    """        
    Parses the request url to create the appropriate response
    """
    # generating the full target path (fp) and its parent (p)
    if request.path=="/" or request.path.startswith('/?'):
        fp = p = workingDir
        base = ""
    elif request.path=="/500":
        # handle error 500
        return {'type': 500}
    else:
        fp = join(workingDir, request.path[1:].split('?')[0])
        p = dirname(fp)
        base = basename(fp)
        print(fp)
    # if the target folder doesn't exist, 404 error
    if not exists(p):
        print("{0} doesn't exist!")
        return {"type": 404}
    if isfile(fp):
        # if the full path corresponds to a file, we check if said file is accessible
        settings = json.load(open(join(workingDir, "settings.json")))
        d = request.path[1:].split('/')[0]
        if not (d in settings['static_dirs'] or d in settings['media_dirs']):
            # if the file is a static or media file, access is forbidden
            return {'type': 403}
        # else we return the bytes content
        return {
            "type": "file",
            "content": open(fp, 'rb').read()
        }
    # if the targer isn't a file we try to load the url pattern
    try:
        urls = json.load(open(
            join(p, "urls.json")
        ))
    except FileNotFoundError:
        # if no pattern is found, throw 404
        return {'type': 404}
    if not base in urls:
        print(base)
        # if the url doesn't correspond to any view, throw 404
        return {'type': 404}
    viewName = urls[base]
    # adding p to sys.path to import the views module
    sys.path.append(
        p
    )
    try:
        import views
        resp = {'type': 'view', 
        'view': None,
        'data': None
        }
        # check if the module contains the target view
        for m in getmembers(views):
            if m[0] == viewName:
                resp['view'] = m[1]
                # get the queries data
                if '?' in request.path:
                    data = parse_qs(
                        request.path.split('?')[1]
                    )
                    for k in data:
                        data[k] = unquote(data[k][0])
                    resp['data'] = data
                return resp
    except ModuleNotFoundError:
        return {'type':404}

src/utils/Server.py
from http.server import HTTPServer, BaseHTTPRequestHandler
from .parse import *
from pathlib import Path
import traceback
from cgi import parse_multipart, parse_header

workingDir = Path().absolute()

class Server(BaseHTTPRequestHandler):
    def do_GET(self):
        # retrieve the request by parsing the URL, and transfer it to the correct view
        request = parseURL(self, workingDir)
        if request['type'] == 404:
            # if the page isn't found, raise a 404 error
            content = "404 page not found".encode('utf-8')
            if exists((path404:=join(workingDir, "404.html"))):
                content = open(path404).read().encode('utf-8')
            self.send_response(404)
        elif request['type'] == 403:
            # forbidden access
            content = "Error 403 forbidden".encode('utf-8')
            self.send_response(403)
        elif request['type'] == 500:
            content = "500 Server error".encode()
            if exists((path500:=join(workingDir, '500.html'))):
                content = open(path500).read().encode('utf-8')
            self.send_response(500)
        elif request['type'] == 'file':
            # if the target is a file, just display the content
            content = request['content']
            self.send_response(200)
        else:
            # if the target is a view, call it by passing the request data as an argument
            content = request['view']({'method': 'GET', 'queries': request['data']}).encode('utf-8')
            self.send_response(200)

        self.end_headers()
        # send the content back to the client
        self.wfile.write(content)
        

    def do_POST(self):
        # retrieve the request by parsing the URL, and transfer it to the correct view
        request = parseURL(self, workingDir)
        # print(request)
        if request['type'] == 404:
            self.send_response(404)
            self.send_header('content-type', 'text/html')
            self.send_header('Location', "/404")
        else:
            # read the content of the request
            length = int(self.headers.get('content-length'))
            ctype, pdict = parse_header(self.headers.get('content-type'))
            if ctype.startswith('multipart/form-data'):
                pdict['boundary'] = pdict['boundary'].encode('ascii')
                data = parse_multipart(self.rfile, pdict)
            else:
                data = parseForm(self.rfile.read(length))
            # transfer the request to the correct view
            try:
                self.send_response(301)
                response = request['view']({'data': data, 'method': 'POST'})
                self.send_header('Location', response)
                self.send_header("content-type", "text/html")
            except Exception:
                # if something goes wrong
                self.send_response(500)
                self.send_header("content-type", "text/html")
                self.send_header("Location", "/500")
                print(traceback.format_exc())

        self.end_headers()